.doom.d/config.org
#+TITLE: Emacs Config
#+PROPERTY: header-args:emacs-lisp :tangle ./config.el :mkdirp yes

* Structure Templates

#+begin_src emacs-lisp
(require 'org-tempo)

(add-to-list 'org-structure-template-alist '("sh" . "src shell"))
(add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
(add-to-list 'org-structure-template-alist '("cpp" . "src C++"))
#+end_src

* User-info
I am writing something for testing.

#+begin_src emacs-lisp
(setq user-full-name  "Almos-Agoston Zediu"
      user-mail-address "zold.almos@gmail.com")
#+end_src

* General fixes
** Keybinds

#+begin_src emacs-lisp
(map! "C-i" 'evil-jump-forward)
(setq-default tab-width 4)

(map! :map makefile-mode-map
      "." 'better-jumper-jump-forward)
#+end_src

** Native compilation

#+begin_src emacs-lisp
(setq package-native-compile t)
#+end_src

** Counsel

#+begin_src emacs-lisp
(use-package counsel
  :custom
  (counsel-linux-app-format-function #'counsel-linux-app-format-function-name-only))
#+end_src

** Override keymaps

#+begin_src emacs-lisp
(require 'bind-key)
(bind-key* "s-l" 'windmove-right)

#+end_src

** Hl-line mode

#+begin_src emacs-lisp
(global-hl-line-mode)
#+end_src

** Reload Emacs Config

#+begin_src emacs-lisp
(defun alm/reload-emacs-config ()
"It relods my config."
(interactive)
  (load "~/.doom.d/config.el"))

(map! :leader "h r c" 'alm/reload-emacs-config)
#+end_src

** Center Buffers

#+begin_src emacs-lisp
(defun alm/visual-fill()
  (setq visual-fill-column-width 100
        visual-fill-column-center-text t
        display-line-numbers nil)
  (visual-fill-column-mode 1))


(add-hook 'org-mode-hook #'alm/visual-fill)
(add-hook 'dired-mode-hook #'alm/visual-fill)
#+end_src

** Scale text

#+begin_src emacs-lisp
(defun alm/scale-text ()
  (text-scale-increase 1))

(add-hook 'org-mode-hook #'alm/scale-text)
(add-hook 'dired-mode-hook #'alm/scale-text)
#+end_src

** Auto-switch to split windows

#+begin_src emacs-lisp
(setq evil-split-window-below t
      evil-vsplit-window-right t)
#+end_src

** Workspace not auto-switching

#+begin_src emacs-lisp
(setq +workspaces-on-switch-project-behavior nil)
#+end_src

* Theme Configuration

#+begin_src emacs-lisp

(load-theme 'doom-challenger-deep t)
(set-frame-parameter (selected-frame) 'alpha '(89 . 89))
(add-to-list 'default-frame-alist '(alpha . (89 . 89)))
(defun load-dark-mode ()
  "It loads my dark configuration."
        (interactive)
        (load-theme 'doom-challenger-deep t)
        (set-frame-parameter (selected-frame) 'alpha '(89 . 75))
        (add-to-list 'default-frame-alist '(alpha . (89 . 75))))

(defun load-light-mode ()
  "It loads my light configuration."
        (interactive)
        (load-theme 'doom-gruvbox-light t)
        (set-frame-parameter (selected-frame) 'alpha '(89 . 89))
        (add-to-list 'default-frame-alist '(alpha . (89 . 89))))

(map! :leader "t m d" 'load-dark-mode)
(map! :leader "t m l" 'load-light-mode)

#+end_src

* Terminal Configuration
** Pop Up terminal

#+begin_src emacs-lisp
(defun terminal ()
  "Initialize or toggle terminal emulator
 If the terminal window is visible hide it.
 If a terminal buffer exists, but is not visible, show it.
 If no terminal buffer exists for the current frame create and show it."
  (interactive)
  (multi-vterm-dedicated-toggle)
  (evil-window-decrease-height 18))
(map! :leader "l" #'terminal)
#+end_src

** Multi-VTerm

#+begin_src emacs-lisp
(map! "s-<return>" 'multi-vterm )
#+end_src

* Projectile Setup

#+begin_src emacs-lisp

#+end_src

* Python setup

#+begin_src emacs-lisp
(use-package! python-black
  :after python)
(add-hook 'python-mode-hook 'python-black-on-save-mode)
(add-hook 'python-mode-hook #'lsp) ; or lsp-deferred
(setq gc-cons-threshold 100000000)
(setq read-process-output-max (* 1024 1024)) ;; 1mb
#+end_src

* JS Setup

#+begin_src emacs-lisp
(add-hook 'js2-mode-hook 'lsp)
#+end_src

* Treemacs setup

#+begin_src emacs-lisp
(require 'treemacs)
(map! :leader "x" 'treemacs)
#+end_src

* Webmode Setup

#+begin_src emacs-lisp
(require 'web-mode)
(add-to-list 'auto-mode-alist '("\\.phtml\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.tpl\\.php\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.[agj]sp\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.as[cp]x\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.erb\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.mustache\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.djhtml\\'" . web-mode))
(setq web-mode-engines-alist
      '(("php"    . "\\.phtml\\'")
        ("blade"  . "\\.blade\\.")
        ("django"   . "\\.html\\."))
)
(setq web-mode-enable-engine-detection t)
#+end_src

* Org Mode

** General settings

#+begin_src emacs-lisp
(require 'org-superstar)
(add-hook 'org-mode-hook (lambda () (org-superstar-mode 1)))
(setq org-directory "~/org/")
(setq org-hide-block-startup t)
(with-eval-after-load 'ox
  (require 'ox-hugo))
(setq org-priority-faces '((65: foreground-color "#660000")
                           (66: foreground-color "#99FFFF")
                           (67: foreground-color "#009150")))
(use-package! org-fancy-priorities
  :ensure t
  :hook (org-mode . org-fancy-priorities-mode)
  :config
  (setq org-fancy-priorities-list '("⚡" "⬆" "⬇" "☕"))
  )

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)
#+end_src

** Automatic tangling
This snippet adds a hook to a =org-mode= buffer so that the config file gets
executed each time such a buffer gets saved.

#+begin_src emacs-lisp
(defun alm/org-babel-tangle-config ()
  (when (string-equal (file-name-directory (buffer-file-name))
                      (expand-file-name "~/.doom.d/"))
    ;; Dynamic scoping to the rescue
    (let ((org-confirm-babel-evaluate nil))
      (org-babel-tangle))))

(add-hook 'org-mode-hook (lambda () (add-hook 'after-save-hook  #'alm/org-babel-tangle-config)))
#+end_src

* CMake Setup

#+begin_src emacs-lisp
(use-package yasnippet
  :config
  (add-to-list 'yas-snippet-dirs "~/.doom.d/snippets/emacs-lisp-mode")
  (add-to-list 'yas-snippet-dirs "~/.doom.d/snippets/emacs-lisp-mode/cmake-mode")
  (yas-global-mode 1))

(defun create-cmake-root(cmake-project-dir-string)
  (let* ((cmake-file-string (concat cmake-project-dir-string "/CMakeLists.txt"))
         (main-cpp-string (concat cmake-project-dir-string "/main.cpp"))
         (build-folder-string (concat cmake-project-dir-string "/build"))
         (debug-folder-string (concat build-folder-string "/Debug"))
         (release-folder-string (concat build-folder-string "/Release")))
        (dired-create-directory build-folder-string)
        (dired-create-directory debug-folder-string)
        (dired-create-directory release-folder-string)
        (with-temp-buffer
        (cmake-mode)
        (yas-minor-mode)
        (yas-expand-snippet (yas-lookup-snippet "cmake_project" 'cmake-mode))
        (when (file-writable-p cmake-file-string)
        (write-region (point-min)
                        (point-max)
                        cmake-file-string))
        (delete-region (point-min)
                       (point-max))
        (cpp-mode)
        (yas-minor-mode)
        (yas-expand-snippet (yas-lookup-snippet "main_cpp" 'cpp-mode))
        (when (file-writable-p main-cpp-string)
        (write-region (point-min)
                        (point-max)
                        main-cpp-string)))
    ))

(defun create-cmake-project (project-root string)
  "Creates a new C++ CMake project"
  (interactive (list (read-directory-name "Select project root: ")
                     (read-string "Name of the project: ")))
  (setq cmake-project-name string)
  (let* ((cmake-project-dir-string (concat project-root string)))
                                  (dired-create-directory cmake-project-dir-string)
                                  (create-cmake-root cmake-project-dir-string)
                                  (projectile-add-known-project cmake-project-dir-string)))

(defun build-cmake-project(mode)
  "Builds a CMake project."
  (let* ((release-mode-string "cmake -S . -B build/ -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=Release && cmake --build build/ && ln -fs build/compile_commands.json")
         (debug-mode-string "cmake -S . -B build/ -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=Debug && cmake --build build/ && ln -fs build/compile_commands.json"))
    (if (equal mode "Debug")
        (message "%s" (concat "Debug mode:\n" (shell-command-to-string debug-mode-string)))
        (message "%s" (concat "Release mode:\n" (shell-command-to-string release-mode-string)))
    )))

(defun build-cmake-project-debug()
  "Builds a CMake project in Debug mode."
  (interactive)
  (build-cmake-project "Debug"))

(defun build-cmake-project-release()
  "Builds a CMake project in Release mode."
  (interactive)
  (build-cmake-project "Release")
  )

(defun run-cmake-project (mode args)
  "Run the CMake project."
  (let* ((status-code-string "; echo \"Process exited with status code: $?\"")
         (release-mode-string (concat "time ./build/Release/" (+workspace-current-name) " " args status-code-string))
         (debug-mode-string (concat "time ./build/Debug/" (+workspace-current-name) " " args status-code-string)))
    (if (equal mode "Debug")
        (message "%s" (concat "Debug mode:\n" (shell-command-to-string debug-mode-string)))
      (message "%s" (concat "Release mode:\n" (shell-command-to-string release-mode-string))))
  ))

(defun run-cmake-project-debug ()
  "Run project in Debug mode."
  (interactive)
  (let* ((args (read-string "Give arguments, if any: ")))
    (run-cmake-project "Debug" args))
  )

(defun run-cmake-project-release ()
  "Run project in Release mode."
  (interactive)
  (let* ((args (read-string "Give arguments, if any: ")))
    (run-cmake-project "Release" args))
  )

(map! :leader :desc "Create a CMake project" "m p" #'create-cmake-project)
(map! :leader :desc "Build CMake project in Release mode." "m r" #'build-cmake-project-release)
(map! :leader :desc "Build CMake project in Debug mode." "m z" #'build-cmake-project-debug)
(map! :leader :desc "Run CMake project in Debug mode." "m Z" #'run-cmake-project-debug)
(map! :leader :desc "Run CMake project in Release mode." "m R" #'run-cmake-project-release)
#+end_src

* VSCode cpptools setup

#+begin_src emacs-lisp
(require 'dap-lldb)
#+end_src

* Email Setup

#+begin_src emacs-lisp
(use-package mu4e
  ;; :load-path "/usr/share/emacs/site-lisp/mu4e/"
  ;; :defer 20 ; Wait until 20 seconds after startup
  :config

  ;; This is set to 't' to avoid mail syncing issues when using mbsync
  (setq mu4e-change-filenames-when-moving t)

  ;; Refresh mail using isync every 10 minutes
  (setq mu4e-update-interval (* 10 60))
  (setq mu4e-get-mail-command "mbsync -a")
  (setq mu4e-root-maildir "~/Mail")

  (setq mu4e-drafts-folder "/[Gmail]/Drafts")
  (setq mu4e-sent-folder   "/[Gmail]/Sent Mail")
  (setq mu4e-refile-folder "/[Gmail]/All Mail")
  (setq mu4e-trash-folder  "/[Gmail]/Trash")
  (setq mu4e-maildir-shortcuts
        '((:maildir "/Inbox"    :key ?i)
        (:maildir "/[Gmail]/Sent Mail" :key ?s)
        (:maildir "/[Gmail]/Trash"     :key ?t)
        (:maildir "/[Gmail]/Drafts"    :key ?d)
        (:maildir "/[Gmail]/All Mail"  :key ?a)))
  (setq smtpmail-smtp-server "smtp.gmail.com"
      smtpmail-smtp-service 465
      smtpmail-stream-type 'ssl)
  (setq message-send-mail-function 'smtpmail-send-it)
  (setq mu4e-compose-signature "Almos Zediu")
)

(use-package org-mime
  :ensure t)
(map! :leader "o m" 'mu4e)
#+end_src

* Dired setup

#+begin_src emacs-lisp
;; (add-hook 'dired-mode-hook #'dired-hide-details-mode)
;; (add-hook 'dired-mode-hook #'all-the-icons-dired-mode)

(use-package dired-hide-details
  :hook (dired-mode . dired-hide-details-mode))

(use-package all-the-icons-dired
  :hook (dired-mode . all-the-icons-dired-mode))

;; (add-to-list 'dired-compress-files-alist '("\\.gz\\'" . "tar $o -r --filesync $i"))


#+end_src

* Agenda Setup
** Org-Super-Agenda config
#+begin_src emacs-lisp
(use-package! org-super-agenda
  :config
  (org-super-agenda-mode 1)
  (setq org-super-agenda-groups
        '(
          (:name "Today"
           :date today
           :time-grid t
           :todo "TODAY")
          (:name "Important"
           :tag "számla"
           :priority "A")
          (:not)))
  )
  :after
  (setq org-super-agenda-header-map nil)

#+end_src

** Basic agenda variables

#+begin_src emacs-lisp
(setq org-agenda-files '("~/Org/agenda.org"
                         "~/Org/orarend.org"
                         "~/Org/projektek"))
(setq org-todo-keywords-for-agenda
      (quote ((sequence "TODO(t)" "NEXT(p)" "WAIT(w)" "CANCELLED" "DONE(r)")
              (sequence "[ ](T)" "[-](S)" "[?](W)" "|" "[X](D)"))))

(setq org-todo-keywords
      (quote ((sequence "TODO(t)" "NEXT(p)" "WAIT(w)" "CANCELLED" "DONE(r)")
              (sequence "[ ](T)" "[-](S)" "[?](W)" "|" "[X](D)"))))

(setq org-capture-templates
      (quote
            (("t" "Personal todo" entry
            (file+headline "~/Org/agenda.org" "Random napi feladatok")
            "* TODO %?\nSCHEDULED: <%(org-read-date)>")
            ("p" "Templates for projects")
            ("pt" "Project-local todo" entry
            (file+headline +org-capture-project-todo-file "Inbox")
            "* TODO %?\n%i\n%a" :prepend t)
            ("pn" "Project-local notes" entry
            (file+headline +org-capture-project-notes-file "Inbox")
            "* %U %?\n%i\n%a" :prepend t)
            ("pc" "Project-local changelog" entry
            (file+headline +org-capture-project-changelog-file "Unreleased")
            "* %U %?\n%i\n%a" :prepend t)
            ("o" "Centralized templates for projects")
            ("ot" "Project todo" entry #'+org-capture-central-project-todo-file "* TODO %?\n %i\n %a" :heading "Tasks" :prepend nil)
            ("on" "Project notes" entry #'+org-capture-central-project-notes-file "* %U %?\n %i\n %a" :heading "Notes" :prepend t)
            ("oc" "Project changelog" entry #'+org-capture-central-project-changelog-file "* %U %?\n %i\n %a" :heading "Changelog" :prepend t))
       ))
#+end_src
* IRC Setup
#+begin_src emacs-lisp
(set-irc-server! "webchat.freenode.net"
  `(:tls t
    :port 6697
    :nick "hrothgar32"
    :sasl-username "hrothgar32"
    :sasl-password "agh54sdE561Q"
    :channels ("#openmw")))
#+end_src


.doom.d/config.el
(require 'org-tempo)

(add-to-list 'org-structure-template-alist '("sh" . "src shell"))
(add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
(add-to-list 'org-structure-template-alist '("cpp" . "src C++"))

(setq user-full-name  "Almos-Agoston Zediu"
      user-mail-address "zold.almos@gmail.com")

(map! "C-i" 'evil-jump-forward)
(setq-default tab-width 4)

(map! :map makefile-mode-map
      "." 'better-jumper-jump-forward)

(setq package-native-compile t)

(use-package counsel
  :custom
  (counsel-linux-app-format-function #'counsel-linux-app-format-function-name-only))

(require 'bind-key)
(bind-key* "s-l" 'windmove-right)

(global-hl-line-mode)

(defun alm/reload-emacs-config ()
"It relods my config."
(interactive)
  (load "~/.doom.d/config.el"))

(map! :leader "h r c" 'alm/reload-emacs-config)

(defun alm/visual-fill()
  (setq visual-fill-column-width 100
        visual-fill-column-center-text t
        display-line-numbers nil)
  (visual-fill-column-mode 1))


(add-hook 'org-mode-hook #'alm/visual-fill)
(add-hook 'dired-mode-hook #'alm/visual-fill)

(defun alm/scale-text ()
  (text-scale-increase 1))

(add-hook 'org-mode-hook #'alm/scale-text)
(add-hook 'dired-mode-hook #'alm/scale-text)

(setq evil-split-window-below t
      evil-vsplit-window-right t)

(setq +workspaces-on-switch-project-behavior nil)

(load-theme 'doom-challenger-deep t)
(set-frame-parameter (selected-frame) 'alpha '(89 . 89))
(add-to-list 'default-frame-alist '(alpha . (89 . 89)))
(defun load-dark-mode ()
  "It loads my dark configuration."
        (interactive)
        (load-theme 'doom-challenger-deep t)
        (set-frame-parameter (selected-frame) 'alpha '(89 . 75))
        (add-to-list 'default-frame-alist '(alpha . (89 . 75))))

(defun load-light-mode ()
  "It loads my light configuration."
        (interactive)
        (load-theme 'doom-gruvbox-light t)
        (set-frame-parameter (selected-frame) 'alpha '(89 . 89))
        (add-to-list 'default-frame-alist '(alpha . (89 . 89))))

(map! :leader "t m d" 'load-dark-mode)
(map! :leader "t m l" 'load-light-mode)

(defun terminal ()
  "Initialize or toggle terminal emulator
 If the terminal window is visible hide it.
 If a terminal buffer exists, but is not visible, show it.
 If no terminal buffer exists for the current frame create and show it."
  (interactive)
  (multi-vterm-dedicated-toggle)
  (evil-window-decrease-height 18))
(map! :leader "l" #'terminal)

(map! "s-<return>" 'multi-vterm )



(use-package! python-black
  :after python)
(add-hook 'python-mode-hook 'python-black-on-save-mode)
(add-hook 'python-mode-hook #'lsp) ; or lsp-deferred
(setq gc-cons-threshold 100000000)
(setq read-process-output-max (* 1024 1024)) ;; 1mb

(add-hook 'js2-mode-hook 'lsp)

(require 'treemacs)
(map! :leader "x" 'treemacs)

(require 'web-mode)
(add-to-list 'auto-mode-alist '("\\.phtml\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.tpl\\.php\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.[agj]sp\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.as[cp]x\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.erb\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.mustache\\'" . web-mode))
(add-to-list 'auto-mode-alist '("\\.djhtml\\'" . web-mode))
(setq web-mode-engines-alist
      '(("php"    . "\\.phtml\\'")
        ("blade"  . "\\.blade\\.")
        ("django"   . "\\.html\\."))
)
(setq web-mode-enable-engine-detection t)

(require 'org-superstar)
(add-hook 'org-mode-hook (lambda () (org-superstar-mode 1)))
(setq org-directory "~/org/")
(setq org-hide-block-startup t)
(with-eval-after-load 'ox
  (require 'ox-hugo))
(setq org-priority-faces '((65: foreground-color "#660000")
                           (66: foreground-color "#99FFFF")
                           (67: foreground-color "#009150")))
(use-package! org-fancy-priorities
  :ensure t
  :hook (org-mode . org-fancy-priorities-mode)
  :config
  (setq org-fancy-priorities-list '("⚡" "⬆" "⬇" "☕"))
  )

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

(defun alm/org-babel-tangle-config ()
  (when (string-equal (file-name-directory (buffer-file-name))
                      (expand-file-name "~/.doom.d/"))
    ;; Dynamic scoping to the rescue
    (let ((org-confirm-babel-evaluate nil))
      (org-babel-tangle))))

(add-hook 'org-mode-hook (lambda () (add-hook 'after-save-hook  #'alm/org-babel-tangle-config)))

(use-package yasnippet
  :config
  (add-to-list 'yas-snippet-dirs "~/.doom.d/snippets/emacs-lisp-mode")
  (add-to-list 'yas-snippet-dirs "~/.doom.d/snippets/emacs-lisp-mode/cmake-mode")
  (yas-global-mode 1))

(defun create-cmake-root(cmake-project-dir-string)
  (let* ((cmake-file-string (concat cmake-project-dir-string "/CMakeLists.txt"))
         (main-cpp-string (concat cmake-project-dir-string "/main.cpp"))
         (build-folder-string (concat cmake-project-dir-string "/build"))
         (debug-folder-string (concat build-folder-string "/Debug"))
         (release-folder-string (concat build-folder-string "/Release")))
        (dired-create-directory build-folder-string)
        (dired-create-directory debug-folder-string)
        (dired-create-directory release-folder-string)
        (with-temp-buffer
        (cmake-mode)
        (yas-minor-mode)
        (yas-expand-snippet (yas-lookup-snippet "cmake_project" 'cmake-mode))
        (when (file-writable-p cmake-file-string)
        (write-region (point-min)
                        (point-max)
                        cmake-file-string))
        (delete-region (point-min)
                       (point-max))
        (cpp-mode)
        (yas-minor-mode)
        (yas-expand-snippet (yas-lookup-snippet "main_cpp" 'cpp-mode))
        (when (file-writable-p main-cpp-string)
        (write-region (point-min)
                        (point-max)
                        main-cpp-string)))
    ))

(defun create-cmake-project (project-root string)
  "Creates a new C++ CMake project"
  (interactive (list (read-directory-name "Select project root: ")
                     (read-string "Name of the project: ")))
  (setq cmake-project-name string)
  (let* ((cmake-project-dir-string (concat project-root string)))
                                  (dired-create-directory cmake-project-dir-string)
                                  (create-cmake-root cmake-project-dir-string)
                                  (projectile-add-known-project cmake-project-dir-string)))

(defun build-cmake-project(mode)
  "Builds a CMake project."
  (let* ((release-mode-string "cmake -S . -B build/ -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=Release && cmake --build build/ && ln -fs build/compile_commands.json")
         (debug-mode-string "cmake -S . -B build/ -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=Debug && cmake --build build/ && ln -fs build/compile_commands.json"))
    (if (equal mode "Debug")
        (message "%s" (concat "Debug mode:\n" (shell-command-to-string debug-mode-string)))
        (message "%s" (concat "Release mode:\n" (shell-command-to-string release-mode-string)))
    )))

(defun build-cmake-project-debug()
  "Builds a CMake project in Debug mode."
  (interactive)
  (build-cmake-project "Debug"))

(defun build-cmake-project-release()
  "Builds a CMake project in Release mode."
  (interactive)
  (build-cmake-project "Release")
  )

(defun run-cmake-project (mode args)
  "Run the CMake project."
  (let* ((status-code-string "; echo \"Process exited with status code: $?\"")
         (release-mode-string (concat "time ./build/Release/" (+workspace-current-name) " " args status-code-string))
         (debug-mode-string (concat "time ./build/Debug/" (+workspace-current-name) " " args status-code-string)))
    (if (equal mode "Debug")
        (message "%s" (concat "Debug mode:\n" (shell-command-to-string debug-mode-string)))
      (message "%s" (concat "Release mode:\n" (shell-command-to-string release-mode-string))))
  ))

(defun run-cmake-project-debug ()
  "Run project in Debug mode."
  (interactive)
  (let* ((args (read-string "Give arguments, if any: ")))
    (run-cmake-project "Debug" args))
  )

(defun run-cmake-project-release ()
  "Run project in Release mode."
  (interactive)
  (let* ((args (read-string "Give arguments, if any: ")))
    (run-cmake-project "Release" args))
  )

(map! :leader :desc "Create a CMake project" "m p" #'create-cmake-project)
(map! :leader :desc "Build CMake project in Release mode." "m r" #'build-cmake-project-release)
(map! :leader :desc "Build CMake project in Debug mode." "m z" #'build-cmake-project-debug)
(map! :leader :desc "Run CMake project in Debug mode." "m Z" #'run-cmake-project-debug)
(map! :leader :desc "Run CMake project in Release mode." "m R" #'run-cmake-project-release)

(require 'dap-lldb)

(use-package mu4e
  ;; :load-path "/usr/share/emacs/site-lisp/mu4e/"
  ;; :defer 20 ; Wait until 20 seconds after startup
  :config

  ;; This is set to 't' to avoid mail syncing issues when using mbsync
  (setq mu4e-change-filenames-when-moving t)

  ;; Refresh mail using isync every 10 minutes
  (setq mu4e-update-interval (* 10 60))
  (setq mu4e-get-mail-command "mbsync -a")
  (setq mu4e-root-maildir "~/Mail")

  (setq mu4e-drafts-folder "/[Gmail]/Drafts")
  (setq mu4e-sent-folder   "/[Gmail]/Sent Mail")
  (setq mu4e-refile-folder "/[Gmail]/All Mail")
  (setq mu4e-trash-folder  "/[Gmail]/Trash")
  (setq mu4e-maildir-shortcuts
        '((:maildir "/Inbox"    :key ?i)
        (:maildir "/[Gmail]/Sent Mail" :key ?s)
        (:maildir "/[Gmail]/Trash"     :key ?t)
        (:maildir "/[Gmail]/Drafts"    :key ?d)
        (:maildir "/[Gmail]/All Mail"  :key ?a)))
  (setq smtpmail-smtp-server "smtp.gmail.com"
      smtpmail-smtp-service 465
      smtpmail-stream-type 'ssl)
  (setq message-send-mail-function 'smtpmail-send-it)
  (setq mu4e-compose-signature "Almos Zediu")
)

(use-package org-mime
  :ensure t)
(map! :leader "o m" 'mu4e)

;; (add-hook 'dired-mode-hook #'dired-hide-details-mode)
;; (add-hook 'dired-mode-hook #'all-the-icons-dired-mode)

(use-package dired-hide-details
  :hook (dired-mode . dired-hide-details-mode))

(use-package all-the-icons-dired
  :hook (dired-mode . all-the-icons-dired-mode))

;; (add-to-list 'dired-compress-files-alist '("\\.gz\\'" . "tar $o -r --filesync $i"))

(use-package! org-super-agenda
  :config
  (org-super-agenda-mode 1)
  (setq org-super-agenda-groups
        '(
          (:name "Today"
           :date today
           :time-grid t
           :todo "TODAY")
          (:name "Important"
           :tag "számla"
           :priority "A")
          (:not)))
  )
  :after
  (setq org-super-agenda-header-map nil)

(setq org-agenda-files '("~/Org/agenda.org"
                         "~/Org/orarend.org"
                         "~/Org/projektek"))
(setq org-todo-keywords-for-agenda
      (quote ((sequence "TODO(t)" "NEXT(p)" "WAIT(w)" "CANCELLED" "DONE(r)")
              (sequence "[ ](T)" "[-](S)" "[?](W)" "|" "[X](D)"))))

(setq org-todo-keywords
      (quote ((sequence "TODO(t)" "NEXT(p)" "WAIT(w)" "CANCELLED" "DONE(r)")
              (sequence "[ ](T)" "[-](S)" "[?](W)" "|" "[X](D)"))))

(setq org-capture-templates
      (quote
            (("t" "Personal todo" entry
            (file+headline "~/Org/agenda.org" "Random napi feladatok")
            "* TODO %?\nSCHEDULED: <%(org-read-date)>")
            ("p" "Templates for projects")
            ("pt" "Project-local todo" entry
            (file+headline +org-capture-project-todo-file "Inbox")
            "* TODO %?\n%i\n%a" :prepend t)
            ("pn" "Project-local notes" entry
            (file+headline +org-capture-project-notes-file "Inbox")
            "* %U %?\n%i\n%a" :prepend t)
            ("pc" "Project-local changelog" entry
            (file+headline +org-capture-project-changelog-file "Unreleased")
            "* %U %?\n%i\n%a" :prepend t)
            ("o" "Centralized templates for projects")
            ("ot" "Project todo" entry #'+org-capture-central-project-todo-file "* TODO %?\n %i\n %a" :heading "Tasks" :prepend nil)
            ("on" "Project notes" entry #'+org-capture-central-project-notes-file "* %U %?\n %i\n %a" :heading "Notes" :prepend t)
            ("oc" "Project changelog" entry #'+org-capture-central-project-changelog-file "* %U %?\n %i\n %a" :heading "Changelog" :prepend t))
       ))

(set-irc-server! "webchat.freenode.net"
  `(:tls t
    :port 6697
    :nick "hrothgar32"
    :sasl-username "hrothgar32"
    :sasl-password "agh54sdE561Q"
    :channels ("#openmw")))
